search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,438 @@
+                                           eks_map_error_to_eks_error (error));
+      return;
+    }
+
+  for (GSList *l = (GSList *) models; l; l = l->next)
+    {
+      EkncContentObjectModel *model = l->data;
+
+      g_variant_builder_open (&builder, G_VARIANT_TYPE_VARDICT);
+
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "title",
+                                                      G_VARIANT_TYPE_STRING);
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "content_type",
+                                                      G_VARIANT_TYPE_STRING);
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "thumbnail_uri",
+                                                      G_VARIANT_TYPE_STRING);
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "ekn_id",
+                                                      G_VARIANT_TYPE_STRING);
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "child_tags",
+                                                      G_VARIANT_TYPE_STRING_ARRAY);
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "tags",
+                                                      G_VARIANT_TYPE_STRING_ARRAY);
+
+      g_variant_builder_close (&builder);
+    }
+
+  shards_strv = strv_from_shard_list ((GSList *) shards);
+
+  eks_article_metadata_complete_query (state->provider->skeleton,
+                                       state->invocation,
+                                       (const gchar * const *) shards_strv,
+                                       g_variant_builder_end (&builder));
+}
+
+static void
+append_construction_prop_from_string (const gchar *key,
+                                      const gchar *str,
+                                      GArray      *values_array,
+                                      GPtrArray   *props_array)
+{
+  guint index = values_array->len;
+  GValue *value = NULL;
+
+  /* Need to be careful to set value *after* the array has been resized */
+  g_array_set_size (values_array, values_array->len + 1);
+  value = &(g_array_index (values_array, GValue, index));
+  g_value_init (value, G_TYPE_STRING);
+  g_value_set_string (value, str);
+  g_ptr_array_add (props_array, g_strdup (key));
```

Comment:
I *believe* that you don't need to copy these strings here and below.

Reply:
I'm afraid not. If I drop the free-func from the gptrarray and don't copy, I get this:

```
(eks-search-provider-v2:3): GLib-GObject-CRITICAL **: g_object_new_is_valid_property: object class 'EkncQueryObject' has no property named '\u0001'

(eks-search-provider-v2:3): GLib-GObject-CRITICAL **: g_object_new_is_valid_property: object class 'EkncQueryObject' has no property named '\u0001'
```

I believe the reason why this happens is that we don't always use static strings when building up the parameter array - in one case we use `g_variant_iter_next` where we have to free the out value (and copy it if it needs to be stored):

```
  g_variant_iter_init (&iter, query_parameters);
  while (g_variant_iter_next (&iter, "{sv}", &iter_key, &iter_value))
    {
      g_autofree gchar *key = iter_key;
      g_autoptr(GVariant) variant = iter_value;

      if (!query_parameter_is_valid (key))
        {
          g_set_error (error,
                       EKS_ERROR,
                       EKS_ERROR_INVALID_REQUEST,
                       "Invalid query parameter: %s",
                       key);
          return NULL;
        }

      append_construction_prop_from_variant (key,
                                             variant,
                                             values_array,
                                             props_array);
    }
```